Add rpc server to helm chart

dstandish opened the pull request:
I call the component "rpc server" beacause the there isn't really just one "internal API" -- the other one is the one used by the webserver.

I think we might want to consider renaming it at the CLI also.

Ideally, also we could somehow mark this (rpc server) private in the helm chart until AIP-44 is "done"

dstandish committed:
Add rpc server to helm chart

dstandish committed:
add tests

dstandish committed:
last test fixes

dstandish committed:
doc fix

dstandish committed:
fix tests

dstandish committed:
remove debug flag

dstandish committed:
fix test

jedcunningham commented on chart/values.yaml:
```diff
@@ -183,6 +183,50 @@ ingress:
     # Http paths to add to the web Ingress after the default path
     succeedingPaths: []
 
+  # Configs for the Ingress of the RPC server Service
```
I don't think we need or want ingress for this component. The service should be sufficient.

jedcunningham commented on helm_tests/webserver/test_rpc_server.py:
```diff
@@ -0,0 +1,923 @@
+class TestRPCServerDeployment:
+    """Tests rpc-server deployment."""
+
+    def test_can_be_disabled(self):
```
We should test it is disabled by default and can be enabled.

jedcunningham commented on helm_tests/webserver/test_rpc_server.py:
```diff
@@ -0,0 +1,923 @@
+
+    @pytest.mark.parametrize("values", [{"config": {"rpcServer": {"base_url": ""}}}, {}])
+    def test_should_not_contain_host_header(self, values):
+        print(values)
```
```suggestion
```

jedcunningham commented on helm_tests/webserver/test_rpc_server.py:
```diff
@@ -0,0 +1,923 @@
+        docs = render_chart(
+            values={
+                "rpcServer": {
+                    "baseUrl": "https://{{ .Release.Name }}.com:21222/mypath/{{ .Release.Name }}/path"
```
hmm, do we need a base url for this component?

dstandish replied:
with webserver it is an airflow config item, which is why the webserver deployment can ref the config value, and it doesn't need to be an item in the component.  but with rpc server, a value is needed for the same reasons (look at the deployment manifest) but, in contrast with webserver, for _rpc server_, there_ is _no_ need for an _airflow_ config for this value. that's why i added it as a component param.  WDYT?

jedcunningham replied:
Forgive me for not being up to speed on 44. If there isn't a config for it, how do the other components know where to send their requests?

dstandish replied:
you're right jed. i think i must have been confused because the config name is `core.internal_api_url` compared with `webserver.baseUrl`

dstandish replied:
and... corrected

jedcunningham commented on helm_tests/webserver/test_rpc_server.py:
```diff
@@ -0,0 +1,923 @@
+            "runAsNonRoot": True,
+        } == jmespath.search("spec.template.spec.securityContext", docs[0])
+
+    def test_rpc_server_security_context_legacy(self):
```
We don't need to support the old style with this new component. If it's too much of a pain with how the helpers are written, maybe we keep it, but lets see if we can whack it.

dstandish replied:
so i figured out that we can disallow it by removing it from the json schema.  no need to mess with the helpers.

now we test that it is forbidden.  but i had to improve the way we invoke the helm command to make that possible, and i split that out into pr https://github.com/apache/airflow/pull/39698

jedcunningham commented on helm_tests/webserver/test_rpc_server.py:
```diff
@@ -0,0 +1,923 @@
+
+        assert expected_ports == jmespath.search("spec.ingress[0].ports", docs[0])
+
+    def test_deprecated_from_param(self):
```
I see deprecated. Can we whack this?

jedcunningham commented on chart/values.yaml:
```diff
@@ -1152,6 +1196,108 @@ migrateDatabaseJob:
   applyCustomEnv: true
 
+rpcServer:
+  enabled: true
```
```suggestion
  enabled: false
```

jedcunningham replied:
You'll have to turn this on for every rpc test case.

jedcunningham commented on helm_tests/airflow_aux/test_basic_helm_chart.py:
```diff
@@ -27,7 +27,7 @@
 from tests.charts.helm_template_generator import render_chart
 
-OBJECT_COUNT_IN_BASIC_DEPLOYMENT = 35
+OBJECT_COUNT_IN_BASIC_DEPLOYMENT = 38
```
With rpc being disabled by default, I don't think we need any of the changes in this file.

jedcunningham commented on helm_tests/security/test_rbac.py:
```diff
@@ -34,13 +34,15 @@
     ("Service", "test-rbac-postgresql"),
     ("Service", "test-rbac-statsd"),
     ("Service", "test-rbac-webserver"),
+    ("Service", "test-rbac-rpc-server"),
```
Or in this one.

jedcunningham reviewed (commented):
Let's prefix with `_`, or maybe `_aip44RpcServer`? I'm hesitant of a docs only change. 

And yes the component wouldn't start, but it would go into crashloop. That said, it seems like we have a miss on our chart integration tests. A component being stuck in crashloop / never actually running should surely be a failure, no?!

dstandish committed:
fix tests

dstandish committed:
fix schema

dstandish committed:
fix helm tests

dstandish committed:
fix more tests

dstandish committed:
rename rpcServer to _rpcServer and add newsfragment / docs re experimental / dev only

dstandish committed:
fixup! rename rpcServer to _rpcServer and add newsfragment / docs re experimental / dev only

dstandish committed:
Print stderr from helm test failures

Previously you would not see the error message; you'd only see that the command failed.  Now you'll see it.

dstandish committed:
fixup! rename rpcServer to _rpcServer and add newsfragment / docs re experimental / dev only

dstandish committed:
fixup! rename rpcServer to _rpcServer and add newsfragment / docs re experimental / dev only

dstandish committed:
disallow legacy securityContext behavior

dstandish committed:
fixup! Print stderr from helm test failures

dstandish committed:
fix test

dstandish committed:
no need for baseUrl

dstandish committed:
remove newsfragment
